language/stdlib/modules/LibraTransactionPublishingOption.move

```diff
@@ -0,0 +1,92 @@
+address 0x1 {
+
+module LibraTransactionPublishingOption {
+    use 0x1::Vector;
+    use 0x1::LibraConfig;
+    use 0x1::LibraTimestamp;
+
+    const SCRIPT_HASH_LENGTH: u64 = 32;
+
+    const ENOT_GENESIS: u64 = 0;
+    const EINVALID_SCRIPT_HASH: u64 = 1;
+
+    /// Defines and holds the publishing policies for the VM. There are three possible configurations:
+    /// 1. No module publishing, only allowlisted scripts are allowed.
+    /// 2. No module publishing, custom scripts are allowed.
+    /// 3. Both module publishing and custom scripts are allowed.
+    /// We represent these as the following resource.
+    struct LibraTransactionPublishingOption {
+        // Only script hashes in the following list can be executed by the network. If the vector is empty, no
+        // limitation would be enforced.
+        script_allow_list: vector<vector<u8>>,
+        // Transactions with sender in this list won't abide to the allow list rules above.
+        sender_allow_list: vector<address>,
+        // Anyone can publish new module if this flag is set to true.
+        module_publishing_allowed: bool,
+    }
+
+    public fun initialize(
+        lr_account: &signer,
+        script_allow_list: vector<vector<u8>>,
+        sender_allow_list: vector<address>,
+        module_publishing_allowed: bool,
+    ) {
+        assert(LibraTimestamp::is_genesis(), ENOT_GENESIS);
+
+        LibraConfig::publish_new_config(
+            lr_account,
+            LibraTransactionPublishingOption {
+                script_allow_list, sender_allow_list, module_publishing_allowed
+            }
+        );
+    }
+
+    // Check if sender can execute script with `hash`
+    public fun is_script_allowed(hash: &vector<u8>, sender: &address): bool {
+        let publish_option = LibraConfig::get<LibraTransactionPublishingOption>();
+
+        Vector::is_empty(&publish_option.script_allow_list)
+            || Vector::contains(&publish_option.script_allow_list, hash)
+            || Vector::contains(&publish_option.sender_allow_list, sender)
+    }
+
+    // Check if a sender can publish a module
+    public fun is_module_allowed(sender: &address): bool {
+        let publish_option = LibraConfig::get<LibraTransactionPublishingOption>();
+
+        publish_option.module_publishing_allowed || Vector::contains(&publish_option.sender_allow_list, sender)
+    }
+
+    // Add `new_hash` to the list of script hashes that is allowed to be executed by the network.
+    public fun add_to_script_allow_list(lr_account: &signer, new_hash: vector<u8>) {
+        // TODO: who is allowed to do this?
```

Comment:
The only option that makes sense here is LibraRoot (and I believe this is even being enforced in `LibraConfig::set`). So I would say add a `Roles::assert_libra_root(lr_account)` here.

Reply:
This is a tricky one. I'm of two minds about this.

- In the long term, I think functions like this should take `&ModifyConfigCapability<Config>` instead of doing a role check
- In the short term, I think doing an explicit role check in all of the functions in this module that call `LibraConfig::set`probably makes sense for consistency. It's redundant, but this is the pattern we use elsewhere...